File: src/NTIX.Core/PackageManager/PackageManagerDetector.cs
using System.Diagnostics;
using System.Linq;
using System.Text.RegularExpressions;
using System.Text.Json;
using System.Threading.Tasks;
using NTIX.Core.Models;

namespace NTIX.Core.PackageManager;

public static class PackageManagerDetector
{

    public static bool IsChocolateyInstalled() => RunProcess("choco --version") != null;
    public static bool IsScoopInstalled() => RunProcess("scoop --version") != null;

    public static async Task<(bool Valid, string? Error, List<string> Warnings)> ValidateManagersAsync(
        NTIXOptions options,
        NTIXConfig config,
        IWingetManager? wingetManager = null)
    {
        var warnings = new List<string>();
        options ??= new NTIXOptions();

        if ((options.Chocolatey?.Enable ?? false) && !IsChocolateyInstalled())
            return (false, "Chocolatey is enabled but not installed. Install from https://chocolatey.org/install", warnings);

        if ((options.Scoop?.Enable ?? false) && !IsScoopInstalled())
            return (false, "Scoop is enabled but not installed. Install from https://scoop.sh", warnings);

        if (options.Winget?.Enable ?? false)
        {
            var mgr = wingetManager ?? new WingetManager();
            if (!mgr.IsInstalled)
            {
                await mgr.EnsureInstalledAsync();
                if (!mgr.IsInstalled)
                    return (false, "Winget is enabled but not installed. Auto-install failed. Install from https://github.com/microsoft/winget-cli", warnings);
            }
        }

        if (config.ChocoPackages.Count > 0 && !(options.Chocolatey?.Enable ?? false))
            warnings.Add("[warn] Chocolatey packages declared but chocolatey not enabled in options");

        if (config.ScoopPackages.Count > 0 && !(options.Scoop?.Enable ?? false))
            warnings.Add("[warn] Scoop packages declared but scoop not enabled in options");

        return (true, null, warnings);
    }

    public static (bool Valid, string? Error, List<string> Warnings) ValidateManagers(
        NTIXOptions options,
        NTIXConfig config)
    {
        var warnings = new List<string>();
        options ??= new NTIXOptions();

        if ((options.Chocolatey?.Enable ?? false) && !IsChocolateyInstalled())
            return (false, "Chocolatey is enabled but not installed. Install from https://chocolatey.org/install", warnings);

        if ((options.Scoop?.Enable ?? false) && !IsScoopInstalled())
            return (false, "Scoop is enabled but not installed. Install from https://scoop.sh", warnings);

        if (config.ChocoPackages.Count > 0 && !(options.Chocolatey?.Enable ?? false))
            warnings.Add("[warn] Chocolatey packages declared but chocolatey not enabled in options");

        if (config.ScoopPackages.Count > 0 && !(options.Scoop?.Enable ?? false))
            warnings.Add("[warn] Scoop packages declared but scoop not enabled in options");

        return (true, null, warnings);
    }

    public static async Task<InstalledPackages> GetInstalledPackagesAsync(Func<IWingetManager>? wingetFactory = null, ICommandRunner? runner = null)
    {
        var cmd = runner ?? new ProcessCommandRunner();
        var result = new InstalledPackages();
        var factory = wingetFactory ?? (() => new WingetManager());
        var wingetManager = factory();

        try
        {
            var wingetPackages = await wingetManager.GetInstalledPackagesAsync();
            foreach (var kvp in wingetPackages)
            {
                result.Winget[kvp.Key] = kvp.Value;
            }
        }
        catch { }

        var chocoOut = await RunProcessAsync(cmd, "choco list -r --local-only --limit-output 2>nul");
        if (!string.IsNullOrEmpty(chocoOut))
        {
            var regex = new Regex(@"^([^|]+)\|([^|]+)\|.*$", RegexOptions.Multiline);
            foreach (Match m in regex.Matches(chocoOut))
            {
                var id = m.Groups[1].Value.Trim();
                var ver = m.Groups[2].Value.Trim();
                if (!string.IsNullOrEmpty(id) && !string.IsNullOrEmpty(ver))
                    result.Chocolatey[id] = ver;
            }
        }

        var scoopOut = await RunProcessAsync(cmd, "scoop list --local-only --limit-output 2>nul");
        if (!string.IsNullOrEmpty(scoopOut))
        {
            var regex = new Regex(@"^([^\s]+)\s+([^\s]+)\s+.*$", RegexOptions.Multiline);
            foreach (Match m in regex.Matches(scoopOut))
            {
                var id = m.Groups[1].Value.Trim();
                var ver = m.Groups[2].Value.Trim();
                if (!string.IsNullOrEmpty(id) && !string.IsNullOrEmpty(ver))
                    result.Scoop[id] = ver;
            }
        }

        return result;
    }

    public static InstalledPackages GetInstalledPackages() => GetInstalledPackagesAsync().GetAwaiter().GetResult();

    public static async Task<Dictionary<string, UpgradeInfo>> GetWingetUpgradablePackagesAsync(Func<IWingetManager>? wingetFactory = null)
    {
        var factory = wingetFactory ?? (() => new WingetManager());
        var wingetManager = factory();
        return await wingetManager.GetUpgradablePackagesAsync();
    }

    public static Dictionary<string, UpgradeInfo> GetWingetUpgradablePackages(Func<IWingetManager>? wingetFactory = null) => GetWingetUpgradablePackagesAsync(wingetFactory).GetAwaiter().GetResult();

    public static async Task<Dictionary<string, UpgradeInfo>> GetChocoUpgradablePackagesAsync(ICommandRunner? runner = null)
    {
        var cmd = runner ?? new ProcessCommandRunner();
        var result = new Dictionary<string, UpgradeInfo>();
        var output = await RunProcessAsync(cmd, "choco outdated --limit-output 2>nul");
        if (string.IsNullOrEmpty(output)) return result;

        var regex = new Regex(@"^([^|]+)\|([^|]+)\|([^|]+)\|.*$", RegexOptions.Multiline);
        foreach (Match m in regex.Matches(output))
        {
            var id = m.Groups[1].Value.Trim();
            var cur = m.Groups[2].Value.Trim();
            var avail = m.Groups[3].Value.Trim();
            if (!string.IsNullOrEmpty(id) && !string.IsNullOrEmpty(cur) && !string.IsNullOrEmpty(avail))
                result[id] = new UpgradeInfo(cur, avail);
        }
        return result;
    }

    public static Dictionary<string, UpgradeInfo> GetChocoUpgradablePackages() => GetChocoUpgradablePackagesAsync().GetAwaiter().GetResult();

    public static async Task<Dictionary<string, UpgradeInfo>> GetScoopUpgradablePackagesAsync(ICommandRunner? runner = null)
    {
        var cmd = runner ?? new ProcessCommandRunner();
        var result = new Dictionary<string, UpgradeInfo>();
        var output = await RunProcessAsync(cmd, "scoop status --json 2>nul");
        if (string.IsNullOrEmpty(output)) return result;

        try
        {
            using var doc = JsonDocument.Parse(output);
            if (doc.RootElement.ValueKind == JsonValueKind.Array)
            {
                foreach (var item in doc.RootElement.EnumerateArray())
                {
                    var id = item.GetProperty("name").GetString();
                    var cur = item.GetProperty("current_version").GetString();
                    var avail = item.GetProperty("latest_version").GetString();
                    if (!string.IsNullOrEmpty(id) && !string.IsNullOrEmpty(cur) && !string.IsNullOrEmpty(avail) && cur != avail)
                        result[id] = new UpgradeInfo(cur, avail);
                }
            }
        }
        catch { }
        return result;
    }

    public static Dictionary<string, UpgradeInfo> GetScoopUpgradablePackages() => GetScoopUpgradablePackagesAsync().GetAwaiter().GetResult();

    public static async Task<Dictionary<string, UpgradeInfo>> GetAllUpgradablePackagesAsync(Func<IWingetManager>? wingetFactory = null, ICommandRunner? runner = null)
    {
        var winget = await GetWingetUpgradablePackagesAsync(wingetFactory);
        var choco = await GetChocoUpgradablePackagesAsync(runner);
        var scoop = await GetScoopUpgradablePackagesAsync(runner);

        var result = new Dictionary<string, UpgradeInfo>();
        foreach (var kvp in winget) result[kvp.Key] = kvp.Value;
        foreach (var kvp in choco) result[kvp.Key] = kvp.Value;
        foreach (var kvp in scoop) result[kvp.Key] = kvp.Value;
        return result;
    }

    public static async Task<bool> ValidateChocoPackageExistsAsync(string id, ICommandRunner? runner = null)
    {
        var cmd = runner ?? new ProcessCommandRunner();
        var output = await RunProcessAsync(cmd, CommandBuilder.BuildChocoSearch(id));
        if (string.IsNullOrEmpty(output)) return false;
        var regex = new Regex($@"^{Regex.Escape(id)}\|", RegexOptions.Multiline | RegexOptions.IgnoreCase);
        return regex.IsMatch(output);
    }

    public static bool ValidateChocoPackageExists(string id)
    {
        var output = RunProcess(CommandBuilder.BuildChocoSearch(id), redirectStderr: true);
        if (string.IsNullOrEmpty(output)) return false;
        var regex = new Regex($@"^{Regex.Escape(id)}\|", RegexOptions.Multiline | RegexOptions.IgnoreCase);
        return regex.IsMatch(output);
    }

    public static async Task<bool> ValidateScoopPackageExistsAsync(string id, ICommandRunner? runner = null)
    {
        var cmd = runner ?? new ProcessCommandRunner();
        var output = await RunProcessAsync(cmd, CommandBuilder.BuildScoopInfo(id));
        if (string.IsNullOrEmpty(output)) return false;
        var pattern = new Regex(@"^\s*Name\s*:", RegexOptions.Multiline | RegexOptions.IgnoreCase);
        return pattern.IsMatch(output);
    }

    public static bool ValidateScoopPackageExists(string id)
    {
        var output = RunProcess(CommandBuilder.BuildScoopInfo(id), redirectStderr: true);
        if (string.IsNullOrEmpty(output)) return false;
        var pattern = new Regex(@"^\s*Name\s*:", RegexOptions.Multiline | RegexOptions.IgnoreCase);
        return pattern.IsMatch(output);
    }

    public static async Task<Dictionary<string, bool?>> ValidateWingetPackagesExistsAsync(
        IEnumerable<string> ids, IWingetManager? wingetManager = null, CancellationToken ct = default)
    {
        var mgr = wingetManager ?? new WingetManager();
        var tasks = ids.Select(async id =>
        {
            try
            {
                var exists = await mgr.PackageExistsAsync(id, ct);
                return (Id: id, Exists: (bool?)exists);
            }
            catch
            {
                return (Id: id, Exists: (bool?)null);
            }
        });
        var results = await Task.WhenAll(tasks);
        return results.ToDictionary(r => r.Id, r => r.Exists);
    }

    public static async Task<Dictionary<string, bool>> ValidateChocoPackagesExistsAsync(
        IEnumerable<string> ids, ICommandRunner? runner = null, CancellationToken ct = default)
    {
        var cmd = runner ?? new ProcessCommandRunner();
        var result = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase);
        await Parallel.ForEachAsync(ids, ct, async (id, token) =>
        {
            var exists = await ValidateChocoPackageExistsAsync(id, cmd);
            result[id] = exists;
        });
        return result;
    }

    public static async Task<Dictionary<string, bool>> ValidateScoopPackagesExistsAsync(
        IEnumerable<string> ids, ICommandRunner? runner = null, CancellationToken ct = default)
    {
        var cmd = runner ?? new ProcessCommandRunner();
        var result = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase);
        await Parallel.ForEachAsync(ids, ct, async (id, token) =>
        {
            var exists = await ValidateScoopPackageExistsAsync(id, cmd);
            result[id] = exists;
        });
        return result;
    }

    internal static string? RunProcess(string cmd, bool redirectStderr = false)
    {
        try
        {
            var psi = new ProcessStartInfo
            {
                FileName = "cmd.exe",
                Arguments = redirectStderr
                    ? $"/c {cmd} 2>&1"
                    : $"/c {cmd} 2>nul",
                RedirectStandardOutput = true,
                UseShellExecute = false,
                CreateNoWindow = true,
                StandardOutputEncoding = System.Text.Encoding.UTF8
            };

            using var p = Process.Start(psi);
            if (p == null) return null;
            var output = p.StandardOutput.ReadToEnd();
            p.WaitForExit();
            return output.Trim();
        }
        catch { return null; }
    }

    private static async Task<string> RunProcessAsync(ICommandRunner runner, string cmd)
    {
        return await runner.RunOutputAsync(cmd, combineStderr: true);
    }
}

File: src/package_manager/package_manager_detector.rs
use std::collections::HashMap;

use crate::models::installed_packages::InstalledPackages;
use crate::models::installed_packages::UpgradeInfo;
use crate::models::{ntix_config::NTIXConfig, options::NTIXOptions};
use crate::package_manager::choco_ops;
use crate::package_manager::command_runner::CommandRunner;
use crate::package_manager::process_command_runner::ProcessCommandRunner;
use crate::package_manager::scoop_ops;
use crate::package_manager::winget_ops;

pub use crate::models::manager_validation::ValidationResult;

pub async fn validate_managers_async(
    options: &NTIXOptions,
    config: &NTIXConfig,
    winget_installed: Option<bool>,
    choco_installed: Option<bool>,
    scoop_installed: Option<bool>,
    runner: Option<&dyn CommandRunner>,
) -> ValidationResult {
    let cmd: &dyn CommandRunner = runner.unwrap_or(&ProcessCommandRunner);
    let (choco_installed, scoop_installed) =
        check_choco_scoop(cmd, choco_installed, scoop_installed).await;

    let winget_installed = match winget_installed {
        Some(v) => v,
        None => {
            let detected = winget_ops::is_installed(cmd).await;
            if options.winget.enable && !detected {
                winget_ops::ensure_installed(cmd).await;
                winget_ops::is_installed(cmd).await
            } else {
                detected
            }
        }
    };

    ValidationResult {
        warnings: collect_warnings(
            options,
            config,
            winget_installed,
            choco_installed,
            scoop_installed,
        ),
        winget_installed,
        choco_installed,
        scoop_installed,
    }
}

async fn check_choco_scoop(
    runner: &dyn CommandRunner,
    choco_override: Option<bool>,
    scoop_override: Option<bool>,
) -> (bool, bool) {
    let choco_installed = match choco_override {
        Some(v) => v,
        None => choco_ops::is_installed(runner).await,
    };
    let scoop_installed = match scoop_override {
        Some(v) => v,
        None => scoop_ops::is_installed(runner).await,
    };
    (choco_installed, scoop_installed)
}

pub fn collect_warnings(
    options: &NTIXOptions,
    config: &NTIXConfig,
    winget_installed: bool,
    choco_installed: bool,
    scoop_installed: bool,
) -> Vec<String> {
    let mut warnings = Vec::new();

    if options.chocolatey.enable && !choco_installed {
        warnings.push(
            "[warn] Chocolatey is enabled but not installed. Skipping Chocolatey packages. Install from https://chocolatey.org/install".to_string(),
        );
    }

    if options.scoop.enable && !scoop_installed {
        warnings.push(
            "[warn] Scoop is enabled but not installed. Skipping Scoop packages and buckets. Install from https://scoop.sh".to_string(),
        );
    }

    if options.winget.enable && !winget_installed {
        warnings.push(
            "[warn] Winget is enabled but not installed. Auto-install failed. Skipping Winget packages. Install from https://github.com/microsoft/winget-cli".to_string(),
        );
    }

    if !config.choco_packages.is_empty() && !options.chocolatey.enable {
        warnings.push(
            "[warn] Chocolatey packages declared but chocolatey not enabled in options".to_string(),
        );
    }

    if !config.scoop_packages.is_empty() && !options.scoop.enable {
        warnings
            .push("[warn] Scoop packages declared but scoop not enabled in options".to_string());
    }

    if !config.winget_packages.is_empty() && !options.winget.enable {
        warnings
            .push("[warn] Winget packages declared but winget not enabled in options".to_string());
    }

    warnings
}

pub async fn get_installed_packages_async(runner: Option<&dyn CommandRunner>) -> InstalledPackages {
    let cmd: &dyn CommandRunner = runner.unwrap_or(&ProcessCommandRunner);

    let mut result = InstalledPackages::default();

    match winget_ops::get_installed_packages(cmd).await {
        Ok(pkgs) => {
            for (id, ver) in pkgs {
                result.winget.insert(id, ver);
            }
        }
        Err(e) => {
            eprintln!("[NTIX] Winget detection failed: {e}");
        }
    }

    let (choco_pkgs, scoop_pkgs) = tokio::join!(
        async {
            match choco_ops::get_installed_packages(cmd).await {
                Ok(pkgs) => pkgs,
                Err(e) => {
                    eprintln!("[NTIX] Chocolatey detection failed: {e}");
                    HashMap::new()
                }
            }
        },
        async {
            match scoop_ops::get_installed_packages(cmd).await {
                Ok(pkgs) => pkgs,
                Err(e) => {
                    eprintln!("[NTIX] Scoop detection failed: {e}");
                    HashMap::new()
                }
            }
        }
    );
    for (id, ver) in choco_pkgs {
        result.chocolatey.insert(id, ver);
    }
    for (id, ver) in scoop_pkgs {
        result.scoop.insert(id, ver);
    }

    result
}

pub async fn get_winget_upgradable_packages_async(
    runner: Option<&dyn CommandRunner>,
) -> HashMap<String, UpgradeInfo> {
    let cmd: &dyn CommandRunner = runner.unwrap_or(&ProcessCommandRunner);
    winget_ops::get_upgradable_packages(cmd)
        .await
        .unwrap_or_default()
}

pub async fn get_choco_upgradable_packages_async(
    runner: Option<&dyn CommandRunner>,
) -> HashMap<String, UpgradeInfo> {
    let cmd: &dyn CommandRunner = runner.unwrap_or(&ProcessCommandRunner);
    choco_ops::get_upgradable_packages(cmd).await
}

pub async fn get_scoop_upgradable_packages_async(
    runner: Option<&dyn CommandRunner>,
) -> HashMap<String, UpgradeInfo> {
    let cmd: &dyn CommandRunner = runner.unwrap_or(&ProcessCommandRunner);
    scoop_ops::get_upgradable_packages(cmd).await
}

pub async fn validate_choco_package_exists_async(
    id: &str,
    runner: &dyn CommandRunner,
) -> Option<bool> {
    choco_ops::package_exists(runner, id).await.ok().flatten()
}

pub async fn validate_scoop_package_exists_async(
    id: &str,
    runner: &dyn CommandRunner,
) -> Option<bool> {
    scoop_ops::package_exists(runner, id).await.ok().flatten()
}

pub async fn validate_winget_packages_exist_async(
    ids: &[String],
    runner: Option<&dyn CommandRunner>,
) -> HashMap<String, Option<bool>> {
    let cmd: &dyn CommandRunner = runner.unwrap_or(&ProcessCommandRunner);

    let tasks = ids.iter().map(|id| async move {
        let exists = winget_ops::package_exists(cmd, id).await.ok();
        (id.clone(), exists)
    });

    futures::future::join_all(tasks).await.into_iter().collect()
}

pub async fn validate_choco_packages_exist_async(
    ids: &[String],
    runner: Option<&dyn CommandRunner>,
) -> HashMap<String, Option<bool>> {
    let cmd: &dyn CommandRunner = runner.unwrap_or(&ProcessCommandRunner);

    let tasks = ids.iter().map(|id| async move {
        let exists = validate_choco_package_exists_async(id, cmd).await;
        (id.clone(), exists)
    });

    futures::future::join_all(tasks).await.into_iter().collect()
}

pub async fn validate_scoop_packages_exist_async(
    ids: &[String],
    runner: Option<&dyn CommandRunner>,
) -> HashMap<String, Option<bool>> {
    let cmd: &dyn CommandRunner = runner.unwrap_or(&ProcessCommandRunner);

    let tasks = ids.iter().map(|id| async move {
        let exists = validate_scoop_package_exists_async(id, cmd).await;
        (id.clone(), exists)
    });

    futures::future::join_all(tasks).await.into_iter().collect()
}
